subsys/dfu/dfu_multi_image/src/dfu_multi_image.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <dfu/dfu_multi_image.h>
+#include <sys/byteorder.h>
+#include <sys/util.h>
+#include <zcbor_decode.h>
+
+#include <errno.h>
+#include <string.h>
+
+#define FIXED_HEADER_SIZE sizeof(uint16_t)
+#define CBOR_HEADER_NESTING_LEVEL 3
+
+struct image_info {
+	int id;
+	size_t size;
+};
+
+struct header {
+	struct image_info images[CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT];
+	size_t image_count;
+};
+
+struct dfu_multi_image_ctx {
+	/* User configuration */
+	uint8_t *buffer;
+	size_t buffer_size;
+	dfu_multi_image_seek_t seek;
+	struct dfu_image_writer writers[CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT];
+	size_t writer_count;
+
+	/* Parsed header */
+	struct header header;
+
+	/* Current parser state */
+	int cur_image_no;
+	size_t cur_offset;
+	size_t cur_item_offset;
+	size_t cur_item_size;
+};
+
+static struct dfu_multi_image_ctx ctx;
+
+static void select_next_image(void)
+{
+	/* TODO: implement logic for skipping MCUboot S0 or S1 image */
+	ctx.cur_image_no++;
+	ctx.cur_item_offset = 0;
+
+	if (ctx.cur_image_no < ctx.header.image_count) {
+		ctx.cur_item_size = ctx.header.images[ctx.cur_image_no].size;
+	} else {
+		ctx.cur_item_size = 0;
+	}
+}
+
+static bool parse_image_info(zcbor_state_t *states, struct image_info *image)
+{
+	bool res;
+
+	res = zcbor_map_start_decode(states);
```

Comment:
Note `return zcbor_pop_error(states);` below. The zcbor library seems to maintain the error code internally and specific operations simply return a boolean. I think such an API was chosen to write a less verbose code. Also, once a parsing operation fails, the following ones won't be executed due to lazy evaluation of boolean expressions in C. So it seems correct, right? If a header is invalid, the error code should be propagated to dfu_multi_image_write() call (I can add a UT for that).

Reply:
ok, I missed that there is logical and operator before next call, so as You said it will return early. OK I understand now, it is passed as function pointer to parse_header, and there it will return false, and it will result in fail, and at the end it will be decoded by the function You mentioned.

Thank You for pointing out my flow in my understanding